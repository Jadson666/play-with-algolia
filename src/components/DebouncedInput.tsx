import { useAtom } from 'jotai'
import debounce from 'lodash/debounce'
import React, { FunctionComponent, useCallback } from 'react'
import { keywordAtom } from '../stores/keywordStore'

interface InputProps {
  onChange?: (e) => void
  time?: number
}

export const DebouncedInput: FunctionComponent<InputProps> = ({
  onChange = () => {},
  time = 0
}) => {
  const [keyword, setKeyword] = useAtom(keywordAtom)
  const debouncedOnChange = debounce(onChange, time)
  const wrappedDebounce = useCallback((e) => debouncedOnChange(e), [])
  const inputOnChange = useCallback((e) => {
    setKeyword(e.target.value)
    wrappedDebounce(e)
  }, [])
  return <input className='searchInput' type='text' onChange={inputOnChange} value={keyword}/>
}
