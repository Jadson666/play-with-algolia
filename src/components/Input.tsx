import debounce from 'lodash/debounce'
import React, { FunctionComponent, useCallback } from 'react'

interface InputProps {
  onChange?: (e) => void
  time?: number
}

export const DebouncedInput: FunctionComponent<InputProps> = ({
  onChange = () => {},
  time = 0
}) => {
  const debouncedOnChange = debounce(onChange, time)
  const inputOnChange = useCallback((e) => debouncedOnChange(e), [])
  return <input className='searchInput' type='text' onChange={inputOnChange} />
}
