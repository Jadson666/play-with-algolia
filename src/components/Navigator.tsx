import Link from 'next/link'
import React, { useMemo } from 'react'
import { useRouter } from 'next/router'
import { getButtonStyle } from '../utils'

export const Navigator = () => {
  const route = useRouter()
  const { searchClass, favoriteClass } = useMemo(() => getButtonStyle(route.pathname), [route.pathname])
  return (
    <NavigatorView searchClass={searchClass} favoriteClass={favoriteClass} />
  )
}

export const NavigatorView = ({ searchClass, favoriteClass }) => {
  return (
    <div className='NavBar'>
      <div style={{ cursor: 'pointer' }}>
        <Link href='/search'>
          <div className={searchClass}>Search</div>
        </Link>
        <Link href='/favorite'>
          <div className={favoriteClass}>Favorite</div>
        </Link>
      </div>
    </div>
  )
}

