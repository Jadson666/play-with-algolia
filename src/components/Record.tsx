import { useAtom } from 'jotai'
import React, { useCallback, useMemo, useState } from 'react'
import { favoriteAddOne, favoriteDelete } from '../stores/events/favorite'
import { favoriteAtom } from '../stores/favoriteStore'

export const Record = ({ data }) => {
  const [{ saved }, setFavorite] = useAtom(favoriteAtom)
  const [hover, setHover] = useState(false)
  const { id, title, author_name, categories } = data

  const handleSave = useCallback((e) => setFavorite(favoriteAddOne(id)), [id])
  const removeFavorite = useCallback(
    (e) => setFavorite(favoriteDelete(id)),
    [id]
  )
  const handleHover = () => setHover(true)
  const handleUnHover = () => setHover(false)
  const isSaved = saved.has(id)

  const button = useMemo(
    () => getButton(hover, isSaved, handleSave, removeFavorite),
    [id, hover, isSaved]
  )
  return (
    <div
      className='record'
      onMouseEnter={handleHover}
      onMouseLeave={handleUnHover}
    >
      <p>{title}</p>
      <div>
        <span>{author_name}</span>
        {categories.map((v) => (
          <span className='tag' key={v}>
            {v}
          </span>
        ))}
      </div>
      {button}
    </div>
  )
}

const getButton = (hover, isSaved, handleSave, handleUnSave) => {
  if (isSaved && hover)
    return (
      <button className='unsave' onClick={handleUnSave}>
        UNSAVE
      </button>
    )
  if (isSaved)
    return (
      <button className='saved' disabled>
        SAVED
      </button>
    )
  if (!isSaved && hover)
    return (
      <button className='save' onClick={handleSave}>
        SAVE
      </button>
    )
  return null
}
