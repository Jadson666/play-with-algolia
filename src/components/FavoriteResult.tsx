import { useAtom } from 'jotai'
import React, { useEffect, useRef } from 'react'
import { apiClient } from '../pages/apiClient'
import { favoriteAtom } from '../stores/favoriteStore'
import { Record } from './Record'

export const FavoriteResults = (props) => {
  const callHappen = useRef(false)
  const [{ saved, isLoading, data }, setFavorite] = useAtom(favoriteAtom)
  const records =
    data.length > 0 ? (
      data.map((v) => <Record key={v.id} data={v} />)
    ) : props.noResult

  const fetch = async () => {
    setFavorite((state) => ({ ...state, isLoading: true }))
    const { hits: result } = await apiClient.search('', {
      filters: getFilterString(saved)
    })
    setFavorite((state) => ({
      ...state,
      isLoading: false,
      data: result as any
    }))
  }

  useEffect(() => {
    if (saved.length === 0) {
      setFavorite((state) => ({
        ...state,
        data: []
      }))
    } else {
      if (!callHappen.current) {
        fetch()
        callHappen.current = true
      }
    }
  }, [saved])

  return (
    <div {...props}>
      {!isLoading && records}
      {isLoading && props.loading}
    </div>
  )
}

const getFilterString = (data) => {
  const temp = data.map(v => `id=${v}`)
  return temp.join(' OR ')
}
