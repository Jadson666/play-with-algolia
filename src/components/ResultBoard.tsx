import React from 'react'
import { Record } from './Record'

export const ResultBoard = ({ noResult, loadingComponent, isLoading, data, ...props }) => {
  const records = data.length > 0 ? data.map((v) => <Record key={v.id} data={v} />) : noResult
  return (
    <div {...props}>
      {!isLoading && records}
      {isLoading && loadingComponent}
    </div>
  )
}
