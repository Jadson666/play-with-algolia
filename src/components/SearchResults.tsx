import { useAtom } from 'jotai'
import React from 'react'
import { searchAtom } from '../stores/searchStore'
import { Record } from './Record'

export const SearchResults = (props) => {
  const [{ isLoading, data }] = useAtom(searchAtom)
  const records =
    data.length > 0
      ? data.map((v) => <Record key={v.id} data={v} />)
      : props.noResult
  return (
    <div {...props}>
      {!isLoading && records}
      {isLoading && props.loading}
    </div>
  )
}
