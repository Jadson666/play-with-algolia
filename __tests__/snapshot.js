import React from 'react'
import renderer from 'react-test-renderer'
import { NavigatorView } from '../src/components/Navigator'

it('renders homepage unchanged', () => {
  const tree = renderer.create(<NavigatorView />).toJSON()
  expect(tree).toMatchSnapshot()
})
