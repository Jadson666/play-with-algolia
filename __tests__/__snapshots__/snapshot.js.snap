// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div
  className="NavBar"
>
  <div
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <div
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Search
    </div>
    <div
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Favorite
    </div>
  </div>
</div>
`;
